File: ruby/056_LuckyBag.rb
# 056 - Lucky Bag（★5）
# 動的計画法, DP法, DP復元

N, S = gets.split.map(&:to_i)
AB = Array.new(N) { gets.split.map(&:to_i) }
dp = Array.new(N + 1) { [false] * (S + 1) }
dp[0][0] = true

N.times do |i|
  (S + 1).times do |j|
    AB[i].each do |x|
      next if j < x
      dp[i + 1][j] = true if dp[i][j - x]
    end
  end
end

if !dp[N][S]
  puts "Impossible"
else
  ans = []
  pos = S
  (N - 1).downto(0) do |i|
    a, b = AB[i]
    if pos >= a && dp[i][pos - a]
      ans[i] = "A"
      pos -= a
    else
      ans[i] = "B"
      pos -= b
    end
  end
  puts ans.join
end

File: kotlin/056_LuckyBag.kt
/** 056 - Lucky Bag（★5）
 * 動的計画法, DP法
 * DP復元
 */

fun readInts() = readLine()!!.split(" ").map(String::toInt)

fun main() {
  val (N, S) = readInts()
  val AB = Array(N) { readInts() }
  var dp = Array(N + 1) { BooleanArray(S + 1) {false} }
  dp[0][0] = true

  for (i in 0 .. N - 1) {
    for (j in 0 .. S) {
      for (x in AB[i]) {
        if (j < x) continue
        if (dp[i][j - x]) dp[i + 1][j] = true
      }
    }
  }

  if (dp[N][S]) {
    var ansRev = StringBuilder()
    var pos = S

    for (i in N - 1 downTo 0) {
      val (a, b) = AB[i]
      if (pos >= a && dp[i][pos - a]) {
        ansRev.append("A")
        pos -= a
      }
      else {
        ansRev.append("B")
        pos -= b
      }
    }
    println(ansRev.reversed())
  } else {
    println("Impossible")
  }
}
